client/blocks/qr-code-login/index.jsx

```diff
@@ -0,0 +1,198 @@
+
+	return false;
+};
+
+const getLoginActionResponse = async ( action, args ) => {
+	const url = new URL( 'https://wordpress.com/wp-login.php' );
+	url.searchParams.append( 'action', action );
+
+	Object.keys( args ).forEach( ( key ) => {
+		url.searchParams.append( key, args[ key ] );
+	} );
+
+	const response = await fetch( url.href );
+	return await response.json();
+};
+
+const fetchQRCodeData = async ( setTokenData, tokenData, anonymousUserId ) => {
+	if ( isStillValidToken( tokenData ) ) {
+		return tokenData;
+	}
+
+	const responseData = await getLoginActionResponse( 'qr-code-token-request-endpoint', {
+		anon_id: anonymousUserId,
+	} );
+
+	setTokenData( responseData.data );
+	setLocalTokenData( responseData.data );
+};
+
+const fetchAuthState = async ( setAuthState, setTokenData, tokenData, anonymousUserId ) => {
+	if ( ! tokenData ) {
+		return;
+	}
+
+	if ( ! isStillValidToken( tokenData ) ) {
+		fetchQRCodeData( setTokenData, tokenData, anonymousUserId );
+		return;
+	}
+
+	const { token, encrypted } = tokenData;
+
+	const responseData = await getLoginActionResponse( 'qr-code-authentication-endpoint', {
+		anon_id: anonymousUserId,
+		token: token,
+		data: encrypted,
+	} );
+
+	setAuthState( responseData.data );
+};
+
+function TokenQRCode( { tokenData } ) {
+	if ( ! tokenData ) {
+		return <QRCodePlaceholder />;
+	}
+	const { token, encrypted } = tokenData;
+	const imageSettings = {
+		src: qrCenter,
+		height: 64,
+		width: 64,
+		excavate: false,
+	};
+	return (
+		<QRCode
+			value={ `https://apps.wordpress.com/get/?campaign=login-qr-code#qr-code-login?token=${ token }&data=${ encrypted }` }
+			size={ 352 }
+			imageSettings={ imageSettings }
+		/>
+	);
+}
+
+function QRCodePlaceholder() {
+	return (
+		<div className="qr-code-login__placeholder">
+			<span className="qr-code-login__corner-box"></span>
+			<span className="qr-code-login__corner-box"></span>
+			<span className="qr-code-login__corner-box"></span>
+		</div>
+	);
+}
+
+function QRCodeLogin() {
+	const translate = useTranslate();
+	const [ tokenData, setTokenData ] = useState( getLocalTokenData() );
+	const [ authState, setAuthState ] = useState( false );
+	const currentTimer = useRef( null );
+
+	const anonymousUserId = getTracksAnonymousUserId();
+
+	// Fetch QR code data.
+	useEffect( () => {
+		fetchQRCodeData( setTokenData, tokenData, anonymousUserId );
+	}, [ setTokenData, tokenData, anonymousUserId ] );
+
+	// Fetch the Auth Data.
+	useEffect( () => {
+		clearInterval( currentTimer.current );
+		currentTimer.current = setInterval( () => {
+			fetchAuthState( setAuthState, setTokenData, tokenData, anonymousUserId );
+		}, AUTH_PULL_INTERVAL );
+
+		return () => clearInterval( currentTimer.current );
+	}, [ setAuthState, setTokenData, tokenData, anonymousUserId, currentTimer ] );
+
+	useEffect( () => {
+		if ( ! authState ) {
+			return;
+		}
+		if ( authState.auth_url ) {
+			window.location.replace( authState.auth_url );
+			return;
+		}
+
+		return null;
```

Comment:
I think we can shorten this a bit:

```suggestion
		if ( authState?.auth_url ) {
			window.location.replace( authState.auth_url );
		}
```